### DISCUSSION TITLE
Compiling the 2020-12 release for cuda

### DISCUSSION THREAD

**Discussion by dosarudaniel** [intel/llvm#3596] [Q&A] [answered]
Hi,

I've been trying to compile the current latest version of this repository but I encountered some problems with the support for this flag: `-fsycl-targets` in adding multiple targets in the list. So I turned my focus on the [latest release](https://github.com/intel/llvm/releases/tag/2020-12) of this repository because I would like to use a stable version of clang++ to compile my sycl code which I will like to run on NVIDIA back end. 

I have followed [these steps](https://github.com/intel/llvm/blob/5eebd1e4bfce1a921409443a3662a1aeabd4ac6e/sycl/doc/GetStartedGuide.md):

`python ~/sycl_workspace/llvm/buildbot/configure.py --cuda`  # see here the full  [output](https://justpaste.it/7z65g) 
`python ~/sycl_workspace/llvm/buildbot/compile.py`  # see here the full [output](https://justpaste.it/1o5vr)   

   
But the compilation failed at this point:
```
[2435/3069] Building CXX object tools/clang/lib/StaticAnalyzer/Frontend/CMakeFiles/obj.clangStaticAnalyzerFrontend.dir/AnalysisConsumer.cpp.o
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "/home/user123/sycl_workspace/llvm/buildbot/compile.py", line 67, in <module>
    ret = main()
  File "/home/user123/sycl_workspace/llvm/buildbot/compile.py", line 63, in main
    return do_compile(args)
  File "/home/user123/sycl_workspace/llvm/buildbot/compile.py", line 40, in do_compile
    subprocess.check_call(cmake_cmd, cwd=abs_obj_dir)
  File "/usr/lib/python2.7/subprocess.py", line 190, in check_call
    raise CalledProcessError(retcode, cmd)
```

**Could you please point me to a fairly stable version of this repository which will compile and support the latest features for SYCL ?**

Thanks!

**Comment by AlexeySachkov** [comment, marked as answer]
Hi @dosarudaniel, 

> Could you please point me to a fairly stable version of this repository which will compile and support the latest features for SYCL ?

Basically, almost each version should be stable enough as we have a number of pre-commit checks including builds and testing with CUDA (see any merged PR) as well as some post-commit checks (here: https://github.com/intel/llvm/actions/workflows/linux_post_commit.yml)

From logs you attached:

> [2404/3069] Performing build step for 'ocl-icd'
> FAILED: tools/sycl/ocl-icd-prefix/src/ocl-icd-stamp/ocl-icd-build lib/libOpenCL.so 
> cd /home/user123/sycl_workspace/llvm/build/tools/sycl/icd_build && /usr/local/bin/cmake -P /home/user123/sycl_workspace/llvm/build/tools/sycl/ocl-icd-prefix/src/ocl-icd-stamp/ocl-icd-build-Release.cmake && /usr/local/bin/cmake -E touch /home/user123/sycl_workspace/llvm/build/tools/sycl/ocl-icd-prefix/src/ocl-icd-stamp/ocl-icd-build
> CMake Error at /home/user123/sycl_workspace/llvm/build/tools/sycl/ocl-icd-prefix/src/ocl-icd-stamp/ocl-icd-build-Release.cmake:49 (message):
>   Command failed: 1
>
>   '/usr/local/bin/cmake' '--build' '.'
>
>  See also
>
>    /home/user123/sycl_workspace/llvm/build/tools/sycl/ocl-icd-prefix/src/ocl-icd-stamp/ocl-icd-build-*.log

It looks like OpenCL ICD build failed, could you please also share those `ocl-icd-build-*.log` files so we can try to analyze what went wrong?

**Reply by bader** [reply]
> Basically, almost each version should be stable enough as we have a number of pre-commit checks including builds and testing with CUDA (see any merged PR) as well as some post-commit checks (here: https://github.com/intel/llvm/actions/workflows/linux_post_commit.yml)

Just to make it clear these post-commit checks: https://github.com/intel/llvm/actions/workflows/linux_post_commit.yml, do not validate CUDA support. 

![image](https://user-images.githubusercontent.com/411965/115693190-6157c380-a368-11eb-8bc5-631346a1041c.png)
`Jenkins/Summary` check runs https://github.com/intel/llvm-test-suite/tree/intel/SYCL tests for DPC++ with CUDA support.

**Reply by dosarudaniel** [reply]
> It looks like OpenCL ICD build failed, could you please also share those `ocl-icd-build-*.log` files so we can try to analyze what went wrong?

Sure, here there are: 
[ocl-icd-build-Release.cmake](https://justpaste.it/5wifg)
[ocl-icd-build-out.log](https://justpaste.it/97jey)
and `ocl-icd-build-err.log` file is empty

**Reply by AlexeySachkov** [reply]
> ocl-icd-build-out.log

I've never encountered this error by myself, but my current guess is that OpenCL-ICD-Loader and OpenCL-Headers sub-projects/dependencies are out of sync: ICD Loader sources reference a macro, which is not defined by Headers anymore:

> /home/user123/sycl_workspace/llvm/build/tools/sycl/OpenCL/icd/test/driver_stub/cl_gl.c:210:64: error: expected declaration specifiers before ‘CL_EXT_SUFFIX__VERSION_1_1’
>                             cl_int *              errcode_ret ) CL_EXT_SUFFIX__VERSION_1_1
>                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~
>/home/user123/sycl_workspace/llvm/build/tools/sycl/OpenCL/icd/test/driver_stub/cl_gl.c:221: error: expected ‘{’ at end of input
>}
> 
>[5/26] Building C object CMakeFiles/OpenCL.dir/loader/icd_dispatch.c.o

Corresponding commits to headers and loader: KhronosGroup/OpenCL-Headers@9d46d9580cc469757ceb8d89842366f6c9a20093, KhronosGroup/OpenCL-ICD-Loader@5f8249691ec8c25775789498951f8e9eb62c201d.

However, I don't have any particular clues why did this happen: based on the configure log, those components were automatically downloaded and I would expect that the latest version is used for both repos (but I need to double-check that)

**Reply by AlexeySachkov** [reply]
Ok I've found the answer:

At 2020-12 tag we used the following versions of headers:

https://github.com/intel/llvm/blob/5eebd1e4bfce1a921409443a3662a1aeabd4ac6e/sycl/CMakeLists.txt#L103

and for ICD:

https://github.com/intel/llvm/blob/5eebd1e4bfce1a921409443a3662a1aeabd4ac6e/sycl/CMakeLists.txt#L147

So, basically you have newer headers and therefore older ICD doesn't compile with them.

I can suggest: manually changing `GIT_TAG` for headers to some older version (something close to ICD tag) or using newer version of intel/llvm

**Reply by dosarudaniel** [reply]
Thank you, @AlexeySachkov !

I've tried that, I am compiling this version of llvm repo: [see my fork here](https://github.com/dosarudaniel/llvm/commit/615058c6f998ee6ce0946b8f133741b07f6ac729). I used this value: [7edca72746c9fc5de5db0acd1fc245cb8ef1b29d](https://github.com/KhronosGroup/OpenCL-Headers/commit/7edca72746c9fc5de5db0acd1fc245cb8ef1b29d) for the GIT_TAG for OpenCL headers.  (Most recent version before December 25th 2020). From the compiling output it seemed that it passed previous point, but now I have another problem related to  SPIRV.:
```
/home/user123/sycl_workspace/llvm/llvm-spirv/lib/SPIRV/libSPIRV/SPIRVEntry.h: In member function ‘SPIRV::SPIRVContinuedInstINTELBase<OC>::EnableIfStruct<OPC, SPIRV::SPIRVType*> SPIRV::SPIRVContinuedInstINTELBase<OC>::getMemberType(size_t) const’:
/home/user123/sycl_workspace/llvm/llvm-spirv/lib/SPIRV/libSPIRV/SPIRVEntry.h:897:70: error: invalid static_cast from type ‘SPIRV::SPIRVEntry*’ to type ‘SPIRV::SPIRVType*’
     return static_cast<SPIRVType *>(SPIRVEntry::getEntry(Elements[I]));
                                                                      ^
[2471/3069] Building CXX object tools/sycl/source/CMakeFiles/sycl_object.dir/detail/builtins_math.cpp.o
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "/home/user123/sycl_workspace/llvm/buildbot/compile.py", line 67, in <module>
    ret = main()
  File "/home/user123/sycl_workspace/llvm/buildbot/compile.py", line 63, in main
    return do_compile(args)
  File "/home/user123/sycl_workspace/llvm/buildbot/compile.py", line 40, in do_compile
    subprocess.check_call(cmake_cmd, cwd=abs_obj_dir)
  File "/usr/lib/python2.7/subprocess.py", line 190, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '--build', '/home/user123/sycl_workspace/llvm/build', '--', 'deploy-sycl-toolchain', 'deploy-opencl-aot', '-j', '32']' returned non-zero exit status 1
```

Again, I am following the same steps:
`python ~/sycl_workspace/llvm/buildbot/configure.py --cuda` # see here the full [output](https://justpaste.it/4ik57)
`python ~/sycl_workspace/llvm/buildbot/compile.py` # see here the full [output](https://justpaste.it/91axh)

Do you have tips on this issue ? Let me know if you need more info.

**Reply by AlexeySachkov** [reply]
> Do you have tips on this issue ? 

@dosarudaniel, yes, it is a known issue which was fixed a week after that release tag was created, you need to apply the following patch to `llvm-spiv` directory: KhronosGroup/SPIRV-LLVM-Translator@cbeeb9b57a3e3ada247d84342c9041c2ee029a25

**Reply by dosarudaniel** [reply]
Thank you @AlexeySachkov , I applied the patch on [my fork](https://github.com/intel/llvm/compare/sycl...dosarudaniel:llvm_latest_release), but now I have another issue:

```
[21/26] Building C object test/loader_test/CMakeFiles/icd_loader_test.dir/test_create_calls.c.o
In file included from /home/user123/sycl_workspace/llvm/build/tools/opencl-aot/OpenCL/icd/test/loader_test/param_struct.h:6,
                 from /home/user123/sycl_workspace/llvm/build/tools/opencl-aot/OpenCL/icd/test/loader_test/test_create_calls.c:9:
/home/user123/sycl_workspace/llvm/build/tools/opencl-aot/OpenCL/inc/CL/cl_gl_ext.h:18:9: note: #pragma message: All OpenGL-related extensions have been moved into cl_gl.h.  Please include cl_gl.h directly.
 #pragma message("All OpenGL-related extensions have been moved into cl_gl.h.  Please include cl_gl.h directly.")
         ^~~~~~~
ninja: build stopped: subcommand failed.
[2545/3069] Building CXX object tools/llvm-spirv/lib/SPIRV/CMakeFiles/LLVMSPIRVLib.dir/libSPIRV/SPIRVType.cpp.o
/home/user123/sycl_workspace/llvm/llvm-spirv/lib/SPIRV/libSPIRV/SPIRVType.cpp: In member function ‘SPIRV::SPIRVWord SPIRV::SPIRVType::getMatrixColumnCount() const’:
/home/user123/sycl_workspace/llvm/llvm-spirv/lib/SPIRV/libSPIRV/SPIRVType.cpp:119:56: warning: type qualifiers ignored on cast result type [-Wignored-qualifiers]
   return static_cast<const SPIRVTypeMatrix *const>(this)->getColumnCount();
                                                        ^
/home/user123/sycl_workspace/llvm/llvm-spirv/lib/SPIRV/libSPIRV/SPIRVType.cpp: In member function ‘SPIRV::SPIRVType* SPIRV::SPIRVType::getMatrixColumnType() const’:
/home/user123/sycl_workspace/llvm/llvm-spirv/lib/SPIRV/libSPIRV/SPIRVType.cpp:124:56: warning: type qualifiers ignored on cast result type [-Wignored-qualifiers]
   return static_cast<const SPIRVTypeMatrix *const>(this)->getColumnType();
                                                        ^
[2557/3069] Building CXX object tools/sycl/source/CMakeFiles/sycl_object.dir/detail/builtins_math.cpp.o
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "/home/user123/sycl_workspace/llvm/buildbot/compile.py", line 67, in <module>
    ret = main()
  File "/home/user123/sycl_workspace/llvm/buildbot/compile.py", line 63, in main
    return do_compile(args)
  File "/home/user123/sycl_workspace/llvm/buildbot/compile.py", line 40, in do_compile
    subprocess.check_call(cmake_cmd, cwd=abs_obj_dir)
  File "/usr/lib/python2.7/subprocess.py", line 190, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '--build', '/home/user123/sycl_workspace/llvm/build', '--', 'deploy-sycl-toolchain', 'deploy-opencl-aot', '-j', '32']' returned non-zero exit status 1
```
`python ~/sycl_workspace/llvm/buildbot/compile.py` # see here the full [output](https://justpaste.it/34n2v) for 3rd trial.

Is this also a known issue?

**Reply by AlexeySachkov** [reply]
> Is this also a known issue?

It is again an issue with OpenCL Loader: two of our sub-components use it and I guess that you need to update `GIT_TAG` in one more cmake file here:

https://github.com/intel/llvm/blob/5eebd1e4bfce1a921409443a3662a1aeabd4ac6e/opencl-aot/CMakeLists.txt#L20

P.S. this situation should be improved in intel/llvm#3485, i.e. we will have a single place where versions of those dependencies are specified

**Reply by dosarudaniel** [reply]
Thank you, @AlexeySachkov, for your help!
I managed to compile it now. 
I used this version of [my llvm fork](https://github.com/dosarudaniel/llvm/tree/5b2e8808b8ead5c02f20521ba42d6d7b8475ef1b).

**Comment by AlexeySachkov** [comment]
> So I turned my focus on the latest release of this repository because I would like to use a stable version of clang++ to compile my sycl code which I will like to run on NVIDIA back end.

BTW, @dosarudaniel, the new release is coming and commit hash is already known, see PR with release notes: https://github.com/intel/llvm/pull/3590/files#diff-34d07c751527855529b6a40917a72f3e46983c6abda27765486594794dd0e19dR3, it is 6a49170027fb - I would assume that newer commit already has all required fixes for issues you have encountered (but it might introduce new ones :))
